Src/Base/AMReX_CuptiTrace.cpp

```diff
@@ -0,0 +1,294 @@
+  uint8_t* buffer = (uint8_t*) malloc(BFR_SIZE + ALIGNMENT);
+  if (buffer == NULL) {
+    amrex::Print() << "Error: CUPTI requested a buffer but memory cannot be allocated\n";
+    exit(EXIT_FAILURE);
+  }
+
+  // Return buffer and buffer size
+  *bfr = ALIGN_BFR(buffer, ALIGNMENT);
+  *size = BFR_SIZE;
+
+  // Controls maximum number of records to be placed in buffer; setting to zero
+  // fills the buffer with as many records as possible, setting to n > 0 fills
+  // with n records before buffer is returned
+  *maxNumRecords = 0;
+}
+
+void CUPTIAPI
+bfrCompleteCallback (CUcontext ctx, uint32_t streamId, uint8_t* bfr,
+		     size_t size, size_t validSize) noexcept
+{ 
+  CUptiResult status;
+  CUpti_Activity* record = NULL;
+
+  if (validSize > 0) {
+    do {
+      status = cuptiActivityGetNextRecord(bfr, validSize, &record);
+      if (status == CUPTI_SUCCESS) {
+	CUpti_Activity_Userdata* recordUserData = new CUpti_Activity_Userdata();
+	CUpti_ActivityKernel4* kernel = (CUpti_ActivityKernel4*) record;
+
+	// Save record data
+	recordUserData->setRecord(record);
+	recordUserData->setStartTime(kernel->start);
+	recordUserData->setEndTime(kernel->end);
+	recordUserData->setTimeElapsed(kernel->end - kernel->start);
+	recordUserData->setStreamID(kernel->streamId);
+	recordUserData->setName(kernel->name);
+	activityRecordUserdata.push_back(recordUserData);
+      }
+      else if (status == CUPTI_ERROR_MAX_LIMIT_REACHED) {
+	// No more records in the buffer
+        break;
+      }
+      else if (status != CUPTI_SUCCESS) {
+        const char* err;
+	cuptiGetResultString(status, &err);
+	amrex::Print() << "Error: " << err <<  "\n";
+	exit(EXIT_FAILURE);
+      }
+    } while (true);
+    
+    size_t dropped;
+    cuptiActivityGetNumDroppedRecords(ctx, streamId, &dropped);
+    if (dropped != 0) {
+      amrex::Print() << (unsigned int) dropped
+		     << " activity records were dropped due to insufficient buffer space\n";
+    }
+  }
+  free(bfr);
+}
+
+void
+initCuptiTrace () noexcept
+{
+  // Enable collection of kernel activity records
+  cuptiActivityEnable(CUPTI_ACTIVITY_KIND_CONCURRENT_KERNEL);
+  
+  // Register callback functions to handle request of buffers to store
+  // activity records and delivery of completed buffers to CUPTI client
+  cuptiActivityRegisterCallbacks(bfrRequestCallback, bfrCompleteCallback);
+
+  amrex::Print() << "CUPTI initialized\n";
+}
+
+void
+cuptiTraceStart () noexcept
+{
+  cudaDeviceSynchronize();
+  cuptiActivityFlushAll(0);  
+  activityRecordUserdata.clear();
+}
+
+void
+cuptiTraceStop () noexcept
+{
+  cudaDeviceSynchronize();
+  cuptiActivityFlushAll(0);
+}
+
+void
+cuptiTraceStop (unsigned boxUintID) noexcept
+{
+  cudaDeviceSynchronize();
+  cuptiActivityFlushAll(0);
+  for (auto record : activityRecordUserdata) {
+    record->setUintID(boxUintID);
+    record->setCharID(("CharID_" + std::to_string(boxUintID)).c_str());
```

Comment:
`("CharID_" + std::to_string(boxUintID))` is a temporary `std::string`.  The pointer returned by `c_str()` will be invalidated when the temporary is gone.  The `charID_` member should be changed to `std::string`.

Reply:
Got it, thanks.  Switched to `std::string`. the only `const char*` in use now is in returning and error before an `amrex::Abort`.